### PR TITLE
Refactor Providers: CloudStorageProvider and NotificationServiceProvider

### PR DESCRIPTION
**Pull request by yokawasa**
# Description

I've made the following refactoring(s)

## CloudStorageProvider

For the future extensibility, I separated provider specific codes from `commands.go` and `gatling_controller.go`, and made it access via unified interface like this:

```go
provider := "aws"
p := cloudstorages.GetProvider(provider)
if p != nil {
  res := (*p).GetXXXX
  ...
}
```

## NotificationServiceProvider

For the future extensibility, I separated provider specific codes from `commands.go` and `gatling_controller.go`, and made it access via unified interface like this:

```go
provider := "slack"
p := notificationservices.GetProvider(provider)
if p != nil {
  err := (*p).Notify(gatlingName, reportURL, secretData)
  ...
}
```

## Utils

Add util functions in a separate package named `utils`, not `controllers`

## Others: Gatling Controller and Relevant test codes

- Modified Gatling Controller so as to use refactored cloud storage providers and notification service providers
- Added and modified test codes accordingly

# Tests

Here is test coverages of gatling operator codes

```bash
make test

?       github.com/st-tech/gatling-operator     [no test files]
?       github.com/st-tech/gatling-operator/api/v1alpha1        [no test files]
ok      github.com/st-tech/gatling-operator/controllers 8.643s  coverage: 33.0% of statements
ok      github.com/st-tech/gatling-operator/controllers/cloudstorages   0.281s  coverage: 100.0% of statements
ok      github.com/st-tech/gatling-operator/controllers/notificationservices    0.487s  coverage: 34.6% of statements
?       github.com/st-tech/gatling-operator/controllers/utils   [no test files]
```

Among them (above), newly added codes are CloudStorageProvider and NotficationServiceProviders. 
- The test codes for CloudStorageProvider covers  100.0% of statements
- The test codes for NotificationServiceProvider covers 34.6% of statements

Regarding the test for NotificationServiceProvider, I confirmed that Slack notification message was sent as expected in my testing environments to check the rest of statements that aren't covered with the test code.

### CODE REVIEW

File: controllers/cloudstorages/cloudstorage.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+	GetCloudStorageReportURL(bucket string, gatlingName string, subDir string) string
+	GetGatlingTransferResultCommand(resultsDirectoryPath string, region string, storagePath string) string
+	GetGatlingAggregateResultCommand(resultsDirectoryPath string, region string, storagePath string) string
+	GetGatlingTransferReportCommand(resultsDirectoryPath string, region string, storagePath string) string
+}
+
+var (
+	cloudStorageProviders = map[string]*CloudStorageProvider{}
+	// use sync.Map to achieve thread safe read and write to map
+	cloudStorageProvidersSyncMap = &sync.Map{}
```
[the remark this answers is not in the document]
Comment (yokawasa) [reply controllers/cloudstorages/cloudstorage.go:19]: Reconciliation function is expected to be concurrent running

[the remark this answers is not in the document]
Comment (tmrekk121) [reply controllers/cloudstorages/cloudstorage.go:19]: Is it necessary for CloudStorageProvidersSyncMap to be a global variable?

[the remark this answers is not in the document]
Comment (yokawasa) [reply controllers/cloudstorages/cloudstorage.go:19]: To follow up more:  i'm using singleton pattern here. I think it's not guaranteed that GetProvider is called from a single caller.  please see also this: https://refactoring.guru/design-patterns/singleton/go/example

**Comment by yokawasa** [comment]
@tmrekk121 any update on your review?

**Review by tmrekk121 — APPROVED** [review]

File: controllers/cloudstorages/aws_test.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+RESULTS_DIR_PATH=testResultsDirectoryPath
+rclone config create s3 s3 env_auth=true region ap-northeast-1
+for source in $(find ${RESULTS_DIR_PATH} -type f -name *.log)
+do
+	rclone copyto ${source} --s3-no-check-bucket --s3-env-auth testStoragePath/${HOSTNAME}.log
+done
+`
+	})
+	Context("Provider is aws", func() {
+		It("provider is aws", func() {
```
Comment (kayamin) [inline controllers/cloudstorages/aws_test.go:94]: @yokawasa 
【nits】
Should test case name be like below? This seems true to other test cases in aws_test.go and gcp_test.go
```suggestion
		It("returns commands with s3 rclone config", func() {
```

Comment (yokawasa) [reply controllers/cloudstorages/aws_test.go]: @kayamin thanks for the suggestion! I took it

**Comment by yokawasa** [comment]
@kayamin based upon your suggestion, I fixed message arg for gingko it func  as well https://github.com/st-tech/gatling-operator/pull/43/commits/bcaab13980d458ba905337c8a9301d294eb7e311

**Review by kayamin — APPROVED** [review]
@yokawasa thank you! LGTM!

**Comment by yokawasa** [comment]
@tmrekk121 @kayamin thanks a lot for the review. I'm merging the PR

**Pull request merged**